platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/WorkNodeCodec.kt

```diff
@@ -55,81 +76,319 @@ typealias IdForNode = (Node) -> Int
-        val idForNode: IdForNode = scheduledNodeIds::getInt
-        nodes.forEach { node ->
-            writeSuccessorReferencesOf(node, idForNode)
-            writeNodeGroup(node.group, idForNode)
+
+        val partialResultsRef = AtomicReference<PersistentList<List<PostExecutionNodes>>>(PersistentList.of())
+
+        // TODO:parallel-cc is this message useful for the context where it may show?
+        runBuildOperations(parallelStore, "saving state nodes" ) {
+            groupedNodes.entries.map { (nodeOwner, groupNodes) ->
+                val groupPath = nodeOwner.path()
+                OperationInfo(displayName = "Storing $groupPath", progressDisplayName = groupPath.path) {
+                    contextSource.writeContextFor(this@writeNodes, groupPath).useToRun {
+                        val postExecutionSuccessors = writeGroupedNodes(nodeOwner, groupNodes, nodeIds)
+                        if (postExecutionSuccessors.isNotEmpty()) {
+                            partialResultsRef.updateAndGet {
+                                it.plus(postExecutionSuccessors)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        val result = IdentityHashMap<ActionNode, List<Node>>()
+        partialResultsRef.get().forEach { nodeAndPostExecutionNodes ->
+            nodeAndPostExecutionNodes.forEach { (node, postExecutionNodes) ->
+                result[node] = postExecutionNodes
+            }
+        }
+        return result
+    }
+
+
+    private
+    fun ReadContext.readNodes(): NodeForId {
+        val nodeIdCount = readSmallInt()
+        val partialResultsRef = AtomicReference<PersistentList<List<NodeId>>>(PersistentList.of())
+        val projectProvider = getSingletonProperty<ProjectProvider>()
+        val groupPaths = readCollectionInto<Path, MutableList<Path>>(::ArrayList) {
+            Path.path(readString())
+        }
+
+        runBuildOperations(parallel = parallelLoad, message = "reading state nodes") {
+            groupPaths.map { groupPath ->
+                OperationInfo(displayName = "Loading $groupPath", progressDisplayName = groupPath.path) {
+                    contextSource.readContextFor(this@readNodes, groupPath).readWith(Unit) {
+                        setSingletonProperty(projectProvider)
+                        val partialResult = readGroupedNodes()
+                        partialResultsRef.updateAndGet {
+                            it.plus(partialResult)
+                        }
+                    }
+                }
+            }
+        }
+        val nodesById = Array<Node?>(nodeIdCount) { null }
+        partialResultsRef.get().forEach { ids ->
+            ids.forEach { (node, id) ->
+                nodesById[id] = node
+            }
+        }
+        return { id: Int -> nodesById[id]!! }
+    }
+
+    private
+    fun <R> IsolateContext.unwrapBuildOperationExceptions(message: String, action: () -> R): R =
```

Comment:
🤔 I'm not completely sold on the idea we should be doing this at this level...

Reply:
Please take another look. Also, see: 

https://github.com/gradle/gradle/pull/30132/files#diff-57a4e23be9128a885df94b1c2a20ab3dec91fc3df3f3f8e21b45c95aed4b255eR23